plugin/pulp/plugin/download.py

```diff
@@ -0,0 +1,122 @@
+DEFAULT_CHUNKSIZE = 1024 * 1024
+
+
+def default_callback_generator(filename=None, size=None, checksums=None):
+    """
+    Creates a closure to write a response to disk, checksum it, and verify
+    its size.
+
+    :param filename: The absolute path the response should be saved to.
+    :param size: The size, in bytes, that the file is expected to be; check is
+                 skipped if this value is not provided.
+    :param checksums: A dictionary of checksums to validate the file with; the
+                      key should be the hash algorithm and the value the
+                      expected hex digest.
+    """
+    # TODO potentially add a way to have this generate the entire checksum set.
+    # Maybe if there's {'algorithm': None}?
+    checksums = checksums or {}
+
+    def callback(self, response):
+        """
+        A callback for the requests_futures library.
+
+        :param resp: The requests response object
+        """
+        hashes = {algorithm: hashlib.new(algorithm) for algorithm in checksums}
+        with open(filename, mode='wb') as fp:
+            chunks = response.iter_content(chunk_size=DEFAULT_CHUNKSIZE)
+            for chunk in chunks:
+                for hash in hashes.values():
+                    hash.update(chunk)
+                fp.write(chunk)
+
+        # Validate that what we downloaded is what the user expects
+        # TODO add more reasonable exceptions
+        if size and size != os.stat(filename).st_size:
+            raise ValueError()
+
+        if checksums != {algorithm: hash.hexdigest() for algorithm, hash in hashes.items()}:
+            raise ValueError()
+
+    return callback
+
+
+class FileFuturesSession(FuturesSession):
+    """
+    This class implements the requests-futures API, with one  small adjuestment. It provides
+    a default callback that writes a file to disk and validates its content. Once it
+    successfully returns, the file is ready to be imported into Pulp.
+
+        >>> future_resp = asynchronous_session.get('https://example.com')
+        >>> filename, resp = future_resp.result()
+        >>> with open(filename) as fd:
+        >>>     django_file = django.core.files.File(fd)
+        >>> artifact = pulp.apps.models.Artifact(file=django_file)
+
+    We should probably provide nice callback options to spit out Artifact objects or even Content
+    objects.
+    """
+
+    def __init__(self, executor=None, max_workers=5, session=None, *args,
+                 **kwargs):
+        super(FileFuturesSession, self).__init__(*args, **kwargs)
+
+    def request(self, *args, **kwargs):
+        """
+        This provides the requests API as well as the requests-futures API.
+
+        The only difference is a few additional keyword arguments are accepted.
+        """
+        callback_generator = kwargs.pop('callback_generator', default_callback_generator)
+        callback_generator_kwargs = kwargs.pop(
```

Comment:
I have a suspicion there is a good reason for the oddity of having a `session` object and also being a `session` which has to do with the "Important" section at the bottom of the README that mentions you must provide a session for Python 3.4 and below, but I didn't investigate further.

>  Then, returns a Future which is a deviation from the requests API. Also, overriding a method and returning a different type is bad.

It is a documented deviation and one that is required to provide a requests-like API in an asynchronous way. Personally, I think this is fine.

As for the problem with the `Queue`, I have no doubt that upstream would be thrilled to receive a contribution that solves the problem. It's general and could be of great use to everyone, so I don't see a good reason to implement it in Pulp-specific code.